Update how we do sdpa testing

drisspg opened the pull request:
## Motivation

This refactor aligns our testing methodology with the Flash Attention upstream repository while addressing several key issues:

1. **Standardized comparison**: We now compare fused kernels against float64 references, using the maximum of a calculated tolerance (based on same-precision math implementation) or standard float32 `atol`.

2. **Reduced redundancy**: Utilizing the same tensors for both same-precision math and fused kernel runs eliminates duplication.

3. **Improved maintainability**: The new approach simplifies tolerance adjustments across all affected tests.

4. **Consistency**: Standardizing tensor comparisons ensures a more uniform and reliable testing suite.

These changes collectively simplify our testing code, improve its maintainability, and provide a more robust framework for validating our attention mechanisms.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #131743

drisspg committed:
update how we do sdpa testing

[ghstack-poisoned]

drisspg committed:
Update on "Update how we do sdpa testing"




[ghstack-poisoned]

drisspg committed:
Update on "Update how we do sdpa testing"




[ghstack-poisoned]

drisspg committed:
Update on "Update how we do sdpa testing"

## Motivation

This refactor aligns our testing methodology with the Flash Attention upstream repository while addressing several key issues:

1. **Standardized comparison**: We now compare fused kernels against float64 references, using the maximum of a calculated tolerance (based on same-precision math implementation) or standard float32 `atol`.

2. **Reduced redundancy**: Utilizing the same tensors for both same-precision math and fused kernel runs eliminates duplication.

3. **Improved maintainability**: The new approach simplifies tolerance adjustments across all affected tests.

4. **Consistency**: Standardizing tensor comparisons ensures a more uniform and reliable testing suite.

These changes collectively simplify our testing code, improve its maintainability, and provide a more robust framework for validating our attention mechanisms.




[ghstack-poisoned]

drisspg committed:
Update

[ghstack-poisoned]

drisspg committed:
Update

[ghstack-poisoned]

drisspg committed:
Update

[ghstack-poisoned]

drisspg committed:
Update

[ghstack-poisoned]

jainapurva reviewed (approved):
Looks good

drisspg committed:
Update

[ghstack-poisoned]

drisspg committed:
Update

[ghstack-poisoned]

jbschlosser commented on test/test_transformers.py:
```diff
@@ -87,37 +87,85 @@ def use_deterministic_algorithims(mode: bool, warn_only: bool):
+    tensor_name: Optional[str] = None
+) -> None:
+    """
+    Compare test tensor against golden and reference tensors.
```
what's the difference between golden and reference? ;p I guess golden is an "ideal reference" (float64 math in this case) and reference is in practice the lp math output

jbschlosser commented on test/test_transformers.py:
```diff
@@ -87,37 +87,85 @@ def use_deterministic_algorithims(mode: bool, warn_only: bool):
+    Compare test tensor against golden and reference tensors.
+    Raises ValueError if compiled error exceeds threshold.
+    """
+    if golden.is_nested and reference.is_nested and test.is_nested:
+        for gold, ref, tst in zip(golden.unbind(), reference.unbind(), test.unbind()):
+            _check_equal(gold, ref, tst, fudge_factor, tensor_name)
+        return
```
will this become unnecessary when NJTs support max? or will NSTs be sent through here

drisspg replied:
hmm I think this would not be necessary for NJT but the tests in this file only run on NST so will still likely be needed

jbschlosser commented on test/test_transformers.py:
```diff
@@ -87,37 +87,85 @@ def use_deterministic_algorithims(mode: bool, warn_only: bool):
-    fudge_factor = fudge_factor if fudge_factor is not None else 1.0
-    atol = get_atol(true_value, computed_value)
-    rtol = get_rtol(true_value, computed_value)
+        grad_checks = [
+            (grad_query_tuple, "Grad_Query"),
+            (grad_key_tuple, "Grad_Key"),
+            (grad_value_tuple, "Grad_Value")
+        ]
```
tiny nit: lol why upper-snake-case here? I think you could avoid the lower() call below if you used lower-snake-case

jbschlosser commented on test/test_transformers.py:
```diff
@@ -3174,46 +3143,28 @@ def get_dropout_mask(output, fused_kernel, batch_size, n_heads, q_len, kv_len, d
-                             atol=grad_k_ref_atol, rtol=grad_k_ref_rtol)
-            self.assertEqual(value.grad, value_ref.grad.to(value.grad.dtype),
-                             atol=grad_v_ref_atol, rtol=grad_v_ref_rtol)
+            dropout_fudge_factor = 1.0 if dropout_p == 0.0 else 1.5
```
conceptually, I feel like you shouldn't compare an output with a reference if there is different dropout structure between the two. ideally either dropout should be shutoff for these reference tests or randomness seeded to produce the same dropout structure. is this part of why the fudge factors seem so high?

drisspg replied:
So the dropout structure is actually exactly the same, I think the main reason for it so that the reduction order is much different between the fused kernels when dropout out is set. That being said let me try removing it and see if there is indeed worse error when dropout is set

jbschlosser commented on test/test_transformers.py:
```diff
@@ -87,37 +87,85 @@ def use_deterministic_algorithims(mode: bool, warn_only: bool):
+    if torch.isnan(test_error).any() and not torch.isnan(ref_error).any():
+        raise ValueError("Output/Grad with NaN")
+
+    # We take the max between a default tolerance set else where for float32 or the calculated fudge_factor numbers
```
not sure what this comment is saying lol

jbschlosser commented on test/test_transformers.py:
```diff
@@ -87,37 +87,85 @@ def use_deterministic_algorithims(mode: bool, warn_only: bool):
+    golden: torch.Tensor,
+    reference: torch.Tensor,
+    test: torch.Tensor,
+    fudge_factor: float,
```
nit: would be nice to have the meaning of the `fudge_factor` documented. I'm assuming it's a multiplicative factor applied to hp math vs. lp math errors. The output under test is allowed to be up to that factor * those errors away from the hp math output.

jbschlosser reviewed (approved):
Looks good - nice changes. still don't understand why low-precision math is closer to high-precision math than the fused backends are though (that's how I'm reading the need for the fudge factor). Is dropout messing with things?

Also test failures indicate that your empirical fudge factors may need to be updated

drisspg committed:
Update on "Update how we do sdpa testing"

## Motivation

This refactor aligns our testing methodology with the Flash Attention upstream repository while addressing several key issues:

1. **Standardized comparison**: We now compare fused kernels against float64 references, using the maximum of a calculated tolerance (based on same-precision math implementation) or standard float32 `atol`.

2. **Reduced redundancy**: Utilizing the same tensors for both same-precision math and fused kernel runs eliminates duplication.

3. **Improved maintainability**: The new approach simplifies tolerance adjustments across all affected tests.

4. **Consistency**: Standardizing tensor comparisons ensures a more uniform and reliable testing suite.

These changes collectively simplify our testing code, improve its maintainability, and provide a more robust framework for validating our attention mechanisms.




[ghstack-poisoned]

drisspg commented:
@pytorchbot merge

drisspg commented:
@pytorchbot merge -f "cant figure it out, I tried running cuda 12.4, I think this gets us in a better state though"
